### DISCUSSION TITLE
cffi::memory new

### DISCUSSION THREAD

**Discussion by nico-robert** [apnadkarni/tcl-cffi#138] [General]
Hello @apnadkarni ,
Me again...
I’m trying to write this function : [D3D11CreateDeviceAndSwapChain](https://docs.microsoft.com/en-us/windows/win32/api/d3d11/nf-d3d11-d3d11createdeviceandswapchain) following this [example](https://github.com/ocornut/imgui/blob/44e3ba115ae70858106ffc8f1c4f827def1aee98/examples/example_win32_directx11/main.cpp#L164)

I know, it's a c++, but I ask a question if I correctly define this [line](https://github.com/ocornut/imgui/blob/44e3ba115ae70858106ffc8f1c4f827def1aee98/examples/example_win32_directx11/main.cpp#L186) with cffi :
`set featureLevelArray [cffi::memory new int\[2\] {0xb000 0xa000} ::d3d11::D3D_FEATURE_LEVEL]` (D3D_FEATURE_LEVEL  is defined like [this](https://docs.microsoft.com/en-us/windows/win32/api/d3dcommon/ne-d3dcommon-d3d_feature_level))

My result of _D3D11CreateDeviceAndSwapChain_ is not `NULL` pointer , but the `out` values yes... He wasn't supposed to have one. So I try to go step by step.

Nicolas

**Comment by apnadkarni** [comment]
Can you paste what your definition of `D3D11CreateDeviceAndSwapChain` looks like?

**Comment by nico-robert** [comment]
like this : 
```
namespace eval ::d3d11 {

    cffi::Wrapper create D3D11 d3d11.dll
    
    cffi::enum define D3D_DRIVER_TYPE_ {
        D3D_DRIVER_TYPE_UNKNOWN   0
        D3D_DRIVER_TYPE_HARDWARE  1
        D3D_DRIVER_TYPE_REFERENCE 2
        D3D_DRIVER_TYPE_NULL      3
        D3D_DRIVER_TYPE_SOFTWARE  4
        D3D_DRIVER_TYPE_WARP      5
    }

    cffi::enum define D3D_FEATURE_LEVEL_ {
        D3D_FEATURE_LEVEL_1_0_CORE 0x1000
        D3D_FEATURE_LEVEL_9_1      0x9100  
        D3D_FEATURE_LEVEL_9_2      0x9200
        D3D_FEATURE_LEVEL_9_3      0x9300
        D3D_FEATURE_LEVEL_10_0     0xa000
        D3D_FEATURE_LEVEL_10_1     0xa100
        D3D_FEATURE_LEVEL_11_0     0xb000
        D3D_FEATURE_LEVEL_11_1     0xb100
        D3D_FEATURE_LEVEL_12_0     0xc000
        D3D_FEATURE_LEVEL_12_1     0xc100
        D3D_FEATURE_LEVEL_12_2     0xc200
    }
    
    cffi::alias define D3D_DRIVER_TYPE   {int {enum D3D_DRIVER_TYPE_}}
    cffi::alias define D3D_FEATURE_LEVEL {int {enum D3D_FEATURE_LEVEL_}}

    D3D11 function D3D11CreateDeviceAndSwapChain pointer.HRESULT {
        pAdapter           {pointer.IDXGIAdapter nullok}
        DriverType         D3D_DRIVER_TYPE
        Software           {pointer.HMODULE nullok}
        Flags              UINT
        pFeatureLevels     pointer.D3D_FEATURE_LEVEL
        FeatureLevels      UINT
        SDKVersion         UINT
        pSwapChainDesc     pointer.::dxgi::DXGI_SWAP_CHAIN_DESC
        ppSwapChain        {pointer.IDXGISwapChain out}
        ppDevice           {pointer.ID3D11Device out}
        pFeatureLevel      {pointer.D3D_FEATURE_LEVEL out}
        ppImmediateContext {pointer.ID3D11DeviceContext out}
    }
}
```
and my definition :
```
set p [cffi::pointer make [winfo id $w] ::win32::HWND]
cffi::pointer safe $p

set createDeviceFlags 0
set D3D11_SDK_VERSION 7

set featureLevelArray [cffi::memory new int\[2\] {0xb000 0xa000} ::d3d11::D3D_FEATURE_LEVEL]

puts "featureLevelArray = $featureLevelArray" ; # if I comment this line D3D11CreateDeviceAndSwapChain command is NULL

set sd [dxgi::DXGI_SWAP_CHAIN_DESC allocate]

dxgi::DXGI_SWAP_CHAIN_DESC setnative $sd BufferCount 2
dxgi::DXGI_SWAP_CHAIN_DESC setnative $sd BufferDesc {
                                                    Width 0 Height 0 Format DXGI_FORMAT_R8G8B8A8_UNORM
                                                    RefreshRate {Numerator 60 Denominator 1}
                                                    ScanlineOrdering 0 Scaling 0
                                                    }


dxgi::DXGI_SWAP_CHAIN_DESC setnative $sd BufferUsage DXGI_USAGE_RENDER_TARGET_OUTPUT
dxgi::DXGI_SWAP_CHAIN_DESC setnative $sd OutputWindow $p
dxgi::DXGI_SWAP_CHAIN_DESC setnative $sd SampleDesc {Count 1 Quality 0}
dxgi::DXGI_SWAP_CHAIN_DESC setnative $sd Windowed   1
dxgi::DXGI_SWAP_CHAIN_DESC setnative $sd SwapEffect DXGI_SWAP_EFFECT_DISCARD


d3d11::D3D11CreateDeviceAndSwapChain NULL \
                              D3D_DRIVER_TYPE_HARDWARE \
                              NULL \
                              $createDeviceFlags \
                              $featureLevelArray \
                              2 \
                              $D3D11_SDK_VERSION $sd g_pSwapChain g_pd3dDevice featureLevel g_pd3dDeviceContext

puts "g_pSwapChain = $g_pSwapChain"
```
I just realized that if I am not in _DEBUG_ mode (print value in my code) the result of the `D3D11CreateDeviceAndSwapChain`  command is `NULL`

**Reply by nico-robert** [reply]
I forgot this field `Flags` in `DXGI_SWAP_CHAIN_DESC `but same result, but now the result is still `NULL`, even in DEBUG mode
`dxgi::DXGI_SWAP_CHAIN_DESC setnative $sd Flags DXGI_SWAP_CHAIN_FLAG_ALLOW_MODE_SWITCH`

**Reply by apnadkarni** [reply]
I'll take a look on the weekend. I'm still rebuilding my development system which bit the dust last week.

**Reply by nico-robert** [reply]
After a day of research, I believe that directX is dead complicated... `IDXGISwapChain` `ID3D11Device`... are structures (I think).Even if I define this function well, I’m not even sure to be able to continue with the next [code](https://github.com/ocornut/imgui/blob/2b1d8e3eafa3204e31aa2de5b42b380d40042930/examples/example_win32_directx11/main.cpp#L202).
Thanks anyway

**Reply by nico-robert** [reply]
I even have the impression that directX it’s more C++ than C

**Reply by nico-robert** [reply]
I’ve made some progress `g_pSwapChain` `g_pd3dDevice` `g_pd3dDeviceContext` are pointers of structures , I finally reworked my `D3D11CreateDeviceAndSwapChain`  function like this :
```
    D3D11 function D3D11CreateDeviceAndSwapChain pointer.HRESULT {
        pAdapter           {pointer.IDXGIAdapter nullok}
        DriverType         D3D_DRIVER_TYPE
        Software           {pointer.HMODULE nullok}
        Flags              UINT
        pFeatureLevels     pointer.D3D_FEATURE_LEVEL
        FeatureLevels      UINT
        SDKVersion         UINT
        pSwapChainDesc     {pointer.::dxgi::DXGI_SWAP_CHAIN_DESC byref}
        ppSwapChain        {pointer.::dxgi::IDXGISwapChainVtbl byref}
        ppDevice           {pointer.ID3D11DeviceVtbl  byref}
        pFeatureLevel      {D3D_FEATURE_LEVEL out}
        ppImmediateContext {pointer.ID3D11DeviceContextVtbl byref}
    }
```
And call this function like this : 
```
set g_pSwapChain        [dxgi::IDXGISwapChainVtbl       allocate]
set g_pd3dDevice        [d3d11::ID3D11DeviceVtbl        allocate]
set g_pd3dDeviceContext [d3d11::ID3D11DeviceContextVtbl allocate]

d3d11::D3D11CreateDeviceAndSwapChain \
                              NULL \
                              D3D_DRIVER_TYPE_HARDWARE \
                              NULL \
                              $createDeviceFlags \
                              $featureLevelArray \
                              2 \
                              $D3D11_SDK_VERSION \
                              $sd \
                              $g_pSwapChain $g_pd3dDevice featureLevel $g_pd3dDeviceContext 
```
Now I’m stuck [here](https://github.com/ocornut/imgui/blob/1dd964f87dc1ea0a76b137baf3401bd201fc2012/examples/example_win32_directx11/main.cpp#L205), I keep on...

Ps: my repo [implottk](https://github.com/nico-robert/implottk) (without directX, only OpenGL)

**Comment by apnadkarni** [comment]
I'm afraid, I've been buried in other stuff and not had time to take a closer look at your work. And as you said, the API is pretty complex so not something I can absorb quickly.

I am a little confused though that you are working with imgui's C++ definitions and not the cimgui C wrappers. Would the latter not be easier? I thought the whole purpose of imgui was that directX and OpenGL are hidden from the application writer. But perhaps I'm wrong on that.

I did tweet about implottk on TclLang.

/Ashok

**Reply by nico-robert** [reply]
For backends there isn't cimgui C wrappers, some libraries backends are easy (OpenGL, glfw) but for DirectX it's more complicated...
Right now, I have no idea which backends I would choose...
As I said in my last message , I'm stuck with this syntax :
```
    ID3D11Texture2D* pBackBuffer;
    g_pSwapChain->GetBuffer(0, IID_PPV_ARGS(&pBackBuffer));
```
For me `g_pSwapChain` is a pointer of structure (I hope) this struct can be find [here](https://github.com/hopk1nz/maf2mp/blob/814cab57dc713d9ff791dfb2a2abeb6af0e2f5a8/Libraries/DXSDK/Include/DXGI.h#L1722) , a little higher in the file you should find the structure in [C++](https://github.com/hopk1nz/maf2mp/blob/814cab57dc713d9ff791dfb2a2abeb6af0e2f5a8/Libraries/DXSDK/Include/DXGI.h#L1666) (I hope...)

`Getbuffer` can be find [here](https://github.com/hopk1nz/maf2mp/blob/814cab57dc713d9ff791dfb2a2abeb6af0e2f5a8/Libraries/DXSDK/Include/DXGI.h#L1781)
Now I would like translate all in `cffi`
my struct : 
```
    cffi::prototype stdcall GetBufferMethod pointer.HRESULT {
        Buffer          UINT
        riid            string ; # I read here and there that this type 'REFIID' could be the uuid of 'IDXGISwapChainVtbl'
        _COM_Outptr_    {pointer out}
    }
    cffi::Struct create IDXGISwapChainVtbl {
        QueryInterface          pointer.HRESULT
        # ...
        GetBuffer               pointer.GetBufferMethod ; # I'm not sure how to define this method , I have chosen  'cffi::prototype stdcall'
        #...
    }
```
my _cffi code_ : 
`set gb [dxgi::IDXGISwapChainVtbl getnative $g_pSwapChain GetBuffer]`
Now I don’t know how to assign this `(0, IID_PPV_ARGS(&pBackBuffer)`
I tried without much conviction  `cffi::call $gb 0 $uuid pBackBuffer` but without success...

> I did tweet about implottk on TclLang.

Thanks for highlighting my work
